### PR TITLE
SPORTS-6 > SPORTS-22 Create an API to create a new tournament

### PR DESCRIPTION
**Pull request by sanjeevkpandit**
- Added APIs:
  - GET `/api/v1/tournaments` - Get a list of all tournaments
  - POST `/api/v1/tournaments` - Create a new tournament
  - GET `/api/v1/tournaments/{id}` - Get a specific tournament info using ID
- Added filter:
  - `Joi` filter for tournament
- Added errors:
  - `Error`
  - `CustomError extends Error`
  - `NotFoundError extends CustomError`
  - `NoRowUpdatedError extends CustomError`

Examples:
```
GET http://localhost:5000/api/v1/tournaments/asdasd
{
    "error": {
        "code": 400,
        "message": "Id must be a number."
    }
}

GET http://localhost:5000/api/v1/tournaments/34
{
    "error": {
        "code": 404,
        "message": "Tournament not Found"
    }
}

GET http://localhost:5000/api/v1/tournaments/1
{
    "data": {
        "id": 1,
        "title": "New title",
        "startDate": "2018-01-21T00:00:00.000Z",
        "finishDate": null,
        "createdAt": "2018-02-05T03:56:36.753Z",
        "updatedAt": "2018-02-05T03:56:36.753Z"
    }
}

```

_**NOTE**: Run the command `npm install` or `yarn` after merging/pulling this._

### CODE REVIEW

File: chess/backend/src/controllers/tournament.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * @export
+ * @param {Request} req
+ * @param {Response} res
+ */
+export async function get(req: Request, res: Response) {
+  const tournamentId = req.params.id;
+  const tournament = await TournamentService.findById(tournamentId);
+
+  res.status(HttpStatus.OK).json({
+    data: tournament
```
Comment (aviskarkc10) [inline chess/backend/src/controllers/tournament.ts]: A question, how are we handling errors?

File: chess/backend/src/controllers/tournament.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * Get list of all tournaments.
+ *
+ * @export
+ * @param {Request} req
+ * @param {Response} res
+ */
+export async function getAll(req: Request, res: Response) {
+  const tournaments = await TournamentService.getAll();
+
+  res.status(HttpStatus.OK).json({
```
Comment (aviskarkc10) [inline chess/backend/src/controllers/tournament.ts]: same here as well?

File: chess/backend/src/controllers/tournament.ts (hunk 1 above)
Comment (cham11ng) [reply chess/backend/src/controllers/tournament.ts]: Nice point. this routing needs `try` `catch` thing. We don't need `async/await` as well

```js
export function get(req: Request, res: Response) {
  const tournamentId = req.params.id;
  TournamentService
    .findById(tournamentId)
    .then((data: TournamentInterface) => {
      res.status(HttpStatus.OK).json({
      data: tournament
    })
    .catch((err: Error) => next());
});
```

Comment (aviskarkc10) [reply chess/backend/src/controllers/tournament.ts]: Async await looks much cleaner than promise. More readable

Comment (aviskarkc10) [reply chess/backend/src/controllers/tournament.ts]:
```javascript
try {
    const tournament = await TournamentService.findById(tournamentId);
    
    res.status(HttpStatus.OK).json({
        data: tournament
    )}
}
catch {
   throw(err);
}
```

**Comment by sanjeevkpandit** [comment]
@aviskarkc10 @cham11ng The description has been updated with new response messages.

File: chess/backend/src/controllers/tournament.ts [hunk 3]
```diff
@@ -1,22 +1,27 @@
...
+    const tournament = await TournamentService.findById(tournamentId);
 
-  res.status(HttpStatus.OK).json({
-    data: tournament
-  });
+    res.status(HttpStatus.OK).json({
+      data: tournament
+    });
+  } catch (error) {
+    next();
```
Comment (cham11ng) [inline chess/backend/src/controllers/tournament.ts]: `next(error)`

File: chess/backend/src/controllers/tournament.ts [hunk 4]
```diff
@@ -25,13 +30,18 @@ export async function get(req: Request, res: Response) {
...
+    const tournaments = await TournamentService.getAll();
 
-  res.status(HttpStatus.OK).json({
-    data: tournaments
-  });
+    res.status(HttpStatus.OK).json({
+      data: tournaments
+    });
+  } catch (error) {
+    next();
```
Comment (cham11ng) [inline chess/backend/src/controllers/tournament.ts]: `next(error)`

File: chess/backend/src/errors/ChessError.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import Error from './Error';
+
+/**
+ * @class ChessError
+ * @extends {Error}
+ */
+class ChessError extends Error {
+  message: string;
+
```
Comment (aviskarkc10) [inline chess/backend/src/errors/ChessError.ts]: why is there extra line after every definition

File: chess/backend/src/errors/Error.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @class Error
+ */
+class Error {
+  message: string;
+
```
Comment (aviskarkc10) [inline chess/backend/src/errors/Error.ts]: extra line here

**Pull request merged**